### PR TITLE
Add a shebang detection script

### PR DESCRIPTION
**Pull request by lenormf**
This script reads the first line of the buffer and does a stupid&simple
association between the last word (excluding interpreter flags) and the
filetype to assign to the buffer.

### CODE REVIEW

**Comment by lenormf** [comment]
Any comments on this ?

File: rc/base/shebang.kak [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    if [ -n "${kak_opt_filetype}" ]; then
+        exit
+    fi
+
+    printf %s "eval -draft -save-regs m %{ try %{
+        exec -save-regs '/' gg <a-x> s^#!\s*[^\n-]+<ret> \;be\"my
+        %sh{
+            filetype=''
+            case \"\${kak_reg_m}\" in
+                zsh|bash|csh|ksh) filetype=sh;;
```
Comment (casimir) [inline rc/base/shebang.kak]: What about `*sh)`? Builtins are almost always the same in shells.

**Comment by casimir** [comment]
I really tried hard to find something but out of unrealistic edge case LGTM.

About the edge case, it doesn't work for a shebang of this form (rarely saw it):
```
#!/usr/bin/env bash -e
```

**Comment by casimir** [comment]
Nevermind I already had a filetype for the buffer I tested in. It works too in this case.

**Comment by lenormf** [comment]
I wouldn't be surprised if someone found a way to break this, it can be tricky sometimes to detect what executable is the interpreter. I hope that it covers at least the most common cases though.

**Comment by casimir** [comment]
@lenormf I was still hoping for a merge. Even more now than the mime-type thing as been statued.

**Comment by maximbaz** [comment]
I was about to start implementing this feature when I found this PR. @lenormf why did you close this? Sounds like a good feature to have, when all other detection mechanisms have failed.

**Comment by lenormf** [comment]
I closed this PR because it's redundant with the `file` functionality. Arguably you could have used it if you don't have `file` installed, but it's probably not worth the time.

**Comment by maximbaz** [comment]
Ah, it seems `file.kak` can recognize variations like `#/bin/bash`, `#!/usr/bin/zsh` and `#!/usr/bin/env bash`, but sadly not `#!/usr/bin/env sh` or `#!/usr/bin/env zsh`. Does anyone know if we can "teach" `file` more shebangs?

**Comment by lenormf** [comment]
Use the `magic` file.

**Comment by codesoap** [comment]
@lenormf could you provide a link to the `magic` file? I couldn't find a`magic.kak` file in `/usr/local/share/kak/rc` or on github.

**Comment by alexherbo2** [comment]
@codesoap https://github.com/mawww/kakoune/issues/2858#issuecomment-482899359

**Comment by alexherbo2** [comment]
@maximbaz Add a `magic` file to your dotfiles.  If you want to take example, you can find mine [here][magic].

[magic]: https://github.com/alexherbo2/configuration/blob/master/home/magic

**Comment by jmpohl** [comment]
@codesoap, were you able to figure out the nuances for OpenBSD? I am on and getting similar behavior to what you described and am not having much success adding `#!/usr/bin/env zsh` to my magic file.

**Comment by jmpohl** [comment]
Disregard previous message. I got my solution and documented it [here](https://github.com/mawww/kakoune/issues/3982#issue-773969548).

**Pull request closed without merging**